x/ecocredit/server/core/msg_server.go

```diff
@@ -0,0 +1,1200 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"github.com/cosmos/cosmos-sdk/orm/types/ormerrors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	basketv1beta1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/basket/v1beta1"
+	marketplacev1beta1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/marketplace/v1beta1"
+	ecocreditv1beta1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1beta1"
+	"github.com/regen-network/regen-ledger/types"
+	"github.com/regen-network/regen-ledger/types/math"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+	"github.com/regen-network/regen-ledger/x/ecocredit/v1beta1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// TODO: Revisit this once we have proper gas fee framework.
+// Tracking issues https://github.com/cosmos/cosmos-sdk/issues/9054, https://github.com/cosmos/cosmos-sdk/discussions/9072
+const gasCostPerIteration = uint64(10)
+
+// CreateClass creates a new class of ecocredit
+//
+// The admin is charged a fee for creating the class. This is controlled by
+// the global parameter CreditClassFee, which can be updated through the
+// governance process.
+func (s serverImpl) CreateClass(ctx context.Context, req *v1beta1.MsgCreateClass) (*v1beta1.MsgCreateClassResponse, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	// Charge the admin a fee to create the credit class
+	adminAddress, err := sdk.AccAddressFromBech32(req.Admin)
+	if err != nil {
+		return nil, err
+	}
+
+	var params ecocredit.Params
+	s.paramSpace.GetParamSet(sdkCtx, &params)
+	if params.AllowlistEnabled && !s.isCreatorAllowListed(sdkCtx, params.AllowedClassCreators, adminAddress) {
+		return nil, sdkerrors.ErrUnauthorized.Wrapf("%s is not allowed to create credit classes", adminAddress.String())
+	}
+
+	err = s.chargeCreditClassFee(sdkCtx, adminAddress)
```

Comment:
we should add a `fee` field to `MsgCreateClass`. Fees should never just happen. let's add a follow-up issue

Reply:
added to follow-ups issue https://github.com/regen-network/regen-ledger/issues/728